src/Analyzers/CSharp/CodeFixes/NewLines/ConditionalExpressionPlacement/ConditionalExpressionPlacementCodeFixProvider.cs

```diff
@@ -0,0 +1,107 @@
+                    CSharpCodeFixesResources.Place_token_on_following_line,
+                    c => UpdateDocumentAsync(document, ImmutableArray.Create(diagnostic), c),
+                    nameof(CSharpCodeFixesResources.Place_token_on_following_line)),
+                context.Diagnostics);
+            return Task.CompletedTask;
+        }
+
+        private static async Task<Document> UpdateDocumentAsync(
+            Document document, ImmutableArray<Diagnostic> diagnostics, CancellationToken cancellationToken)
+        {
+            var text = await document.GetTextAsync(cancellationToken).ConfigureAwait(false);
+            var root = await document.GetRequiredSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+            using var _ = ArrayBuilder<TextChange>.GetInstance(out var edits);
+
+            foreach (var diagnostic in diagnostics)
+            {
+                var questionToken = root.FindToken(diagnostic.Location.SourceSpan.Start);
+                Contract.ThrowIfTrue(questionToken.Kind() != SyntaxKind.QuestionToken);
+
+                var conditional = (ConditionalExpressionSyntax)questionToken.GetRequiredParent();
+
+                AddEdits(text, conditional.QuestionToken, conditional.WhenTrue, edits);
+                AddEdits(text, conditional.ColonToken, conditional.WhenFalse, edits);
+            }
+
+            var changedText = text.WithChanges(edits);
+            return document.WithText(changedText);
+        }
+
+        private static void AddEdits(
+            SourceText text,
+            SyntaxToken token,
+            ExpressionSyntax nextExpression,
+            ArrayBuilder<TextChange> edits)
+        {
+            // Cases to consider
+            // x ?
+            // x ? 
+            // x ? /* comment */
+            // x /* comment */ ?
+            // x /* comment */ ? /* comment */
+            //
+            // in all these cases, we want to grab the question, and any spaces that follow and remove that, but we
+            // leave the rest where it is. We then move the question right before the start of the next token.  The
+            // same logic applies to the colon token.
+
+            var start = token.SpanStart;
+            var end = token.Span.End;
+
+            while (end < text.Length && text[end] == ' ')
+                end++;
+
+            if (end < text.Length && SyntaxFacts.IsNewLine(text[end]))
+            {
+                while (start > 0 && text[start - 1] == ' ')
+                    start--;
+            }
+
+            edits.Add(new TextChange(TextSpan.FromBounds(start, end), ""));
+            edits.Add(new TextChange(new TextSpan(nextExpression.SpanStart, 0), token.Text + " "));
+        }
+
+        public override FixAllProvider? GetFixAllProvider()
+            => FixAllProvider.Create(async (context, document, diagnostics) => await UpdateDocumentAsync(document, diagnostics, context.CancellationToken).ConfigureAwait(false));
```

Comment:
I mean, it is a batch fixer :)

Reply:
So why not just `WellKnownFixAllProviders.BatchFixer`?

Reply:
batch-fixer is outright terrible.  It makes all the changes, then does a diff, then tries to reconcile the diff edits.  However, diff edits are non-deterministic.  It's dependent on whatever the diff engine computes the change as (including replacing the entire construct with a new construct).  This makes conflicts on edits very possible (and we have several bugs on this for real fixers).  The approach here is minimal and sound, with no chance of conflicts.

Reply:
> and we have several bugs on this for real fixers

Could you link an example? Interested to know more about this.

`BatchFixer` is very commonly used (in roslyn, roslyn-analyzers, aspnetcore, runtime, and other 3rd party libraries). It would be good to understand when it's suitable and when it's not. If scenarios where it's not suitable are somewhat common, it would be very helpful that this is documented in https://learn.microsoft.com/en-us/dotnet/csharp/roslyn-sdk/tutorials/how-to-write-csharp-analyzer-code-fix

Reply:
It is never suitable for any case where multiple edits may be needed in the same file.  Fundamentally, the design is not sound (for the reason i listed above).  Teams can certainly use it, and cross their fingers for the best.  But if you want ensured correctness, it must be avoided.